httpbp/client_middlewares.go

```diff
@@ -0,0 +1,194 @@
+// to fail based on a configurable failure ratio based on total failures and
+// requests.
+func CircuitBreaker(config Config) ClientMiddleware {
+	pool := &sync.Pool{
+		New: func() interface{} {
+			return &gobreaker.Settings{
+				Interval: 60 * time.Second,
+				ReadyToTrip: func(counts gobreaker.Counts) bool {
+					failureRatio := float64(counts.TotalFailures) / float64(counts.Requests)
+					return counts.Requests >= config.CircuitBreaker.MinRequests && failureRatio >= config.CircuitBreaker.MinFailureRatio
+				},
+			}
+		},
+	}
+	breakers := &sync.Map{}
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (*http.Response, error) {
+			host := req.URL.Hostname()
+
+			// new circuit breakers should rarely get allocated: the pool reduces
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				pool.Put(newBreaker)
+			}
+
+			resp, err := breaker.(*gobreaker.CircuitBreaker).Execute(func() (interface{}, error) {
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					err = resp.Body.Close()
+					if err != nil {
+						return nil, err
+					}
+					return nil, err
+				}
+				return resp, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+			return resp.(*http.Response), nil
+		})
+	}
+}
+
+func applyMiddlewares(transport http.RoundTripper, middlewares ...ClientMiddleware) http.RoundTripper {
+	// add middlewares in reverse so the first in the list is the outermost
+	for i := len(middlewares) - 1; i >= 0; i-- {
+		transport = middlewares[i](transport)
+	}
+	return transport
+}
+
+// Retries provides a retry middleware.
+func Retries(config retry.Config) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			retrybp.Do(req.Context(), func() error {
+				resp, err = next.RoundTrip(req)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, retry.Attempts(1))
+			if err != nil {
+				return nil, err
+			}
+			return resp, nil
+		})
+	}
+}
+
+// LoadBalancer implements a round-robin load balancer.
+func LoadBalancer(config Config) ClientMiddleware {
+	attempts := 0
+	offset := uint64(0)
+	hosts := config.LoadBalancer.Hosts
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+
+			remaining := attempts
+			for remaining > 0 {
+				i := atomic.AddUint64(&offset, 1)
+				hostIndex := i % uint64(len(hosts))
+				req.Host = hosts[hostIndex]
+				req.URL.Host = hosts[hostIndex]
+
+				resp, err := next.RoundTrip(req)
+				if err == gobreaker.ErrOpenState || err == gobreaker.ErrTooManyRequests {
+					remaining--
```

Comment:
so the current logic is that, if the host picked by loadbalancer has circuit breaker open, we try the previous one instead, and if the first one still has loadbalancer open, we try it again and return the error (line 147). this means that in the case of circuit breaker is open, we actually put more pressure on the first host.

I think this logic needs more work, and since it looks like the requirement for load balancer is a quite unique requirement that's not really needed by most services, I'd suggest we just remove this from the pr, and y'all can just implement it in your code base instead.

Reply:
Agreed, it's very specific and depends on what the user wants to accomplish with it.